### PR TITLE
Nailing down automatic circuit cutter's API

### PR DESCRIPTION
**Pull request by zeyueN**
**Context:**
When passing a tape to the automatic circuit cutter (KaHyPar), several hyperparameters (e.g. `k`, `imbalance`, etc.) need to be supplied. Depending on use cases, sensible default values should be derived based on known information (which usually comes at different points in time) such as the available devices and the circuits. However, power users should also be able to override and supply their own parameters. We thus must provide a unified interface that's general enough to allow these types of interactions.

Drawing similarities between distributing circuit fragments to devices and the classical problem of distributing computation tasks dags to multiple workers, the constraints coming from devices should be separated from the requirements coming from the computations. This separation is realized inside the added class `CutStrategy`, whose attributes contain constraints from devices and whose method `get_cut_kwargs()` takes circuit requirements at runtime and returns complete sets of cutting hyperparameters. In terms of UI, both the device constrains and the calling of `get_cut_kwargs()` could be done either manually by users or within decorators. `dataclass` was elected to make the implementation of the init function cleaner.

**Description of the Change:**
Adds:
- a class `CutStrategy` that can be used as an interface to coordinate devices info, user constraints, and circuit requirements into  partition hyperparameters ready for partitioning methods.

**Possible Drawbacks:**
Gate cut compatibility not fully explored

### CODE REVIEW

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:241]: So would these be the user-facing `kwargs` that can be optionally passed, e.g., in the `cut_circuit()` signature?

**Review by trbromley — COMMENTED** [review]
Thanks @zeyueN!

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:227]: Exactly, kahypar has the functionality to optionally pass a list of max fragment sizes if there's such a constraint. However since partition nodes corresponds to gates, it's only applicable to `fragment_gates` rather than `fragment_wires`. For `fragment_wires` to achieve similar functionality, I don't have a good idea right now other than sampling a bunch of partition trials and pick the one that satisfies the condition, if there's any. Or, we could remove `fragment_wires` for now, though it's realistically the more useful one.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:251]: Good point! I think this can safely default to `num_tape_gates` which is basically to impose no gate constraint. `min_device_gates`, when not supplied, defaults to `max_device_gates` which is to say there's only one device size.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:258]: The extreme cases here are solely for the calculation of `k`'s bounds and has nothing to do with how we run them. It doesn't mean we will carry out the execution, nor does it guarantee the petitioner will come up with fragments to realize these extreme scenarios. In other words, it's the best/worst scenarios in terms of number of fragments required for running a tape on a pool of devices.

Take your example of devices with 4, 5, and 6 wires:
- if the circuit has 4 wires, `k_lower` and `k_upper` both defaults to 1 since any device suffices;
- if the circuit has 5 wires, `k_lower` defaults to 1 since it can be run on the 5- or 6-wire devices, but `k_upper` defaults to 2 since, if a user requires, two 4-wire devices will be required. Of course `k` can be higher than `k_upper` (e.g. running a (2,2,1) split on three 4-wire devices), but that's for the user to decide if they want to explore the option.
- if the circuit has 11 wires, you'll need at least two 6- or 5- wire devices, but three 4-wire devices. So `k_lower` defaults to 2 since this is the least number of cuts we'll need in order to fit things onto the available devices. `k_upper` defaults to 3 in the scenario we only use the 4-wire one. But users get to decide if they want to do more exploration by setting it higher than 3. For example, if the circuit is in fact itself already completely disconnected in to 4 parts.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:263]: The comment above perhaps can help a bit. But the general idea is that, since users don't usually know what k to supply, we need to figure out some sort of sensible defaults. But since we don't know what k is the best before we do any partitioning, we need to try a range of ks and compare later, so what's required is then a sensible default of this range of ks, which is `k_upper` and `k_lower`.

Now `k_lower` is very easy to find out, if your largest device is 40 wires, you'll at least to cut a 100-wire circuit into 3 parts.

But `k_upper` is slightly trickier since there's no strict bound and one can try to cut a circuit into as many fragments as possible. But for a sensible default value, I thought it could be the number of splits needed as if we only use the smallest device in the pool. Or another way to think about this, especially for the case of only one device, this is the number of splits required if we choose to only PARTIALLY occupy wires on the device.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:274]: Since we are trying a range of k, and each kahypar call takes only a single k, we need a list of things we want to try. 

But we do have some freedom in how we do this. In its current state, each `CutConfig` corresponds to a single call to kahypar with fixed k and precalculated imbalance (depending on k), etc. An alternative is to generate them dynamically, which will require moving all the arguments and some logic of the helper function into the class. But let's not over-engineer this 😆

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:415]: Ah mixed mode, good point. Yes the whole class will change if we want to support that. Should be doable.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:410]: Oh it's the original circuit's gate/wire number, just to keep an record of it since some properties' derivations may need those.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:413]: It'll be however we want to represent the edge, so yes for `nx` edges it'll be Ant's pair of gates with wire data. Alternatively we can also keep things lean in `raw_cuts` by keeping only kahypar edge indices, and use a `@property` that maps things back to nx edges.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:241]: Yes, though the name perhaps is not the most insightful for users without reading the docstring, especially for `k_lower` and `k_upper`. One idea is to call `k_lower` just `k` in `cut_circuit()`, but that can also be confusing lol.

**Comment by trbromley** [comment]
[sc-14152]

**Comment by zeyueN** [comment]
Edited the description. Promoting to real PR.

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:454]:
```suggestion
        min_free_gates (int): Minimum allowed circuit depth for the shallowest available device.
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:455]:
```suggestion
            Optional, defaults to ``max_free_gates``.
```
(not sure, but think this was renamed)

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:473]:
```suggestion
        tape_dag = qml.transforms.tape_to_graph(tape)
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:489]: Nice to see this mapped out! However, since it's blocked by the `cut_circuit` transform, how about this example goes into the "Usage details" section of `cut_circuit` and for the examples in this function we focus just on a simple use case of `CutStrategy`, e.g.,

```rst
    The following cut strategy specifies that a circuit should be cut into between
    ``2`` to ``5`` fragments, with each fragment having at most ``6`` wires and
    at least ``4`` wires:

    >>> cut_strategy = qcut.CutStrategy(
    ...     max_free_wires=6,
    ...     min_free_wires=4,
    ...     num_fragments_probed=(2, 5),
    ... )
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:501]: `imbalance_tolerance` is missing from the docstrings

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:582]: Would be good to move the example below args and returns

Comment (trbromley) [inline pennylane/transforms/qcut.py:582]: Perhaps this section is the only bit we need.

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:596]: I think you might have answered in previous comments, but why a list? Also it seems at odds with the example:
```
        >>> cut_kwargs = cut_strategy.get_cut_kwargs(tape_dag)
        >>> cut_kwargs.update({'extra_param': 0})
        >>> my_partitioner(tape_dag, **cut_kwargs)
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:639]: Are asserts the best choice for user validation? It's more length by `if-raise` seems more natural.

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:655]:
```suggestion
        for the graph partitioner.
```

**Review by trbromley — COMMENTED** [review]
Thanks @zeyueN! I left a few suggestions/comments, but overall this looks great and happy to approve once these are resolved.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:684]: 👍 done

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py]: The original idea is to also allow a non-consecutive list for the probed `k`s. But at the end I opted to interpret lists only as range bounds in this initial implementation because I'm not entirely sure if non-consecutive list of `k`s are really needed; and I want to leave some room in the future for such changes.

As for removing the `int` option, user would then have to provide say `(2, 3)` in order to pick just `2`, which is slightly cumbersome.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py]: The returned list contains all potential cut kwargs to explore, i.e. different `k`s and `imbalance`s. You are right the example is wrong, have fixed it.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py]: switched.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:765]: This is me wanting to be slightly more forgiving before checking if one of `devices` and `max_free_wires` is provided, where if neither are provided but `min_free_wires` are, then I interpret `max_free_wires` to be the same as `min_free_wires`. Unlikely situation, but if it happens this interpretation should be safe.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py]: Because the length implicitly signifies the number of fragments, i.e. `k`.

**Comment by zeyueN** [comment]
@trbromley I think I've addressed all the comments and ready for review.

The documentation check workflow is failing for some reason. Any idea how to fix?

File: pennylane/transforms/qcut.py
Comment (anthayes92) [reply pennylane/transforms/qcut.py:759]: Thanks this makes sense, though now I'm not sure I understand the docstring where for `max_free_wires` we have "`Optional only when ``devices`` is provided` and vice versa for `devices`". Maybe I'm misunderstanding the use of `Optional` but this seems to suggest that at least one of the two must be passed?

Comment (zeyueN) [reply pennylane/transforms/qcut.py:927]: You are correct that the docstring doesn't accurately describe this behaviour. This was intentional since this is sort of a "failsafe" behaviour which doesn't need to be advertised to users since it adds nothing functionally new and could potentially confuse people even more. In the end, I do want to force users to provide either `devices` or `max_free_wires`.

File: doc/releases/changelog-dev.md
Comment (trbromley) [inline doc/releases/changelog-dev.md:84]:
```suggestion
  A class `CutStrategy` which acts as an interface and coordinates device/user
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:20]:
```suggestion
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:910]:
```suggestion
    devices: InitVar[Union[qml.Device, Sequence[qml.Device]]] = None
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:876]:
```suggestion
            Optional, defaults to unlimited imbalance. Used only if there's a known hard balancing
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:894]:
```suggestion
```
Is this still needed?

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:950]: Is this the right error message? We raise an error if:
- Devices is not a list of tuple, or
- Any element in device is not a `qml.Device`

Maybe the error message should just read:
```
Argument `devices` must be a list or tuple containing elements of type qml.Device
```
(Also, should we just check for `Sequence` rather than list/tuple?)

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:992]:
```suggestion
        Deriving kwargs for a given circuit and feeding them to a custom partitioner, along with
        extra parameters specified using ``extra_kwargs``:
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:916]: Since the docs shows these attributes without docstrings:
image

Would it be possible to add quick one line explanations? E.g.,
```
    imbalance_tolerance: float = None
    #: The global maximum allowed imbalance for all partition trials
```
(I'm not completely sure on the UI to get this to show up in the docs so might need some trial and error)

**Review by trbromley — COMMENTED** [review]
Thanks @zeyueN, just a few more comments.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:894]: I initially deleted it but then thought it could add some context. But now looking at it again I guess the text desrcription already did a fairly good job at that. So ya let's delete it.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py:936]: Ya definitely screwed up that sentence, switched to your suggestion.

File: pennylane/transforms/qcut.py
Comment (zeyueN) [reply pennylane/transforms/qcut.py]: Done, the comment should be above the line.
image

Comment (trbromley) [reply pennylane/transforms/qcut.py]: Awesome, thanks for figuring out the right orientation for the comment!

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:890]:
```suggestion
```

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:994]:
```suggestion
            partitioner method.
```
Just a minor formatting issue:
image

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:957]:
```suggestion
                    f"`qml.Device`"
```
I like the added context but I'm not sure if it'll be confusing when failing the `(not isinstance(d, qml.Device) for d in devices)` branch.

E.g., they pass `devices = [qml.device("default.qubit", wires=2), "not_a_dev"]`. This will fail because of `"not_a_dev"` but the raise message would say `got <class 'list'>` which might confuse the user as to what the problem is.

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1148]:
```suggestion
                    f"The attempted number of fragments seems high with lower bound at {k_lower}."
```
Only because I'm not sure a user can really cancel other than killing the job.

File: pennylane/transforms/qcut.py
Comment (trbromley) [inline pennylane/transforms/qcut.py:1156]:
```suggestion
                    f"The numer of partition attempts seems high ({len(ks)})."
```

**Review by trbromley — APPROVED** [review]
Awesome, looks great - thanks @zeyueN! I left some final comments, but otherwise good to go! Approved :green_circle:

**Pull request merged**
